cuML/src/common/allocatorAdapter.hpp

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright (c) 2018-2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "../../../ml-prims/src/utils.h"
+
+#include "../cuML.hpp"
+
+namespace ML {
+
+/**
+ * @code{.cpp}
+ * void foo( const cumlHandle_impl& h, ... , cudaStream_t stream )
+ * {
+ *     auto execution_policy = ML::exec_policy(h.getDeviceAllocator(),stream);
+ *     thrust::for_each(execution_policy->on(stream), ... );
+ * }
+ * @endcode
+ */
+class thrustAllocatorAdapter
+{
+public:
+    using value_type = char;
+
+    thrustAllocatorAdapter() = delete;
+
+    thrustAllocatorAdapter(std::shared_ptr<deviceAllocator> allocator, cudaStream_t stream)
+        : _allocator(allocator), _stream(stream)
+    {}
+    
+    ~thrustAllocatorAdapter() {}
+    
+    char* allocate(const size_t size)
+    {
+        return static_cast<char*>(_allocator->allocate( size, _stream ));
+    }
+
+    void deallocate(char* ptr, const size_t size)
+    {
+        _allocator->deallocate( ptr, size, _stream );
+    }
+
+private:
+    std::shared_ptr<deviceAllocator>    _allocator;
+    cudaStream_t                        _stream = 0;
+};
+
+namespace 
+{
+    thrustAllocatorAdapter _decltypeHelper{0,0};
+}
+
+/**
+ * @brief Returns a unique_ptr to a Thrust CUDA execution policy that uses the
+ * passed in allocator for temporary memory allocation.
+ *
+ * @Param allocator The allocator to use
+ * @Param stream    The stream that the allocator will use
+ *
+ * @Returns A Thrust execution policy that will use allocator for temporary memory
+ * allocation.
+ */
+inline auto exec_policy(std::shared_ptr<deviceAllocator> allocator, cudaStream_t stream) -> std::unique_ptr<decltype(thrust::cuda::par(_decltypeHelper)),std::function<void(decltype(thrust::cuda::par(_decltypeHelper))*)> >
+{
+    thrustAllocatorAdapter * alloc{nullptr};
+
+    alloc = new thrustAllocatorAdapter(allocator, stream);
+
+    using T = decltype(thrust::cuda::par(*alloc));
+
+    auto deleter = [alloc](T* pointer) {
+        delete alloc;
+        delete pointer;
+    };
+    
+    std::unique_ptr<T, decltype(deleter)> policy{new T(*alloc), deleter};
+    return policy;
+}
+
```

Comment:
I see your point, but I am not convinced that we should add this function. Regarding passing in `const cumlHandle_impl& h` instead of `std::shared_ptr<deviceAllocator> allocator` does my argument that the execution policy does not depend on `cumlHandle_impl` made above convince you?

Reply:
I don't see an argument about this above.  Above what?

Reply:
Sorry I assumed you a reading top to bottom, propably github decided to hide part of the conversation. I was refering this reply made above:

Reply begin:

Regarding passing in ML::cumHandle_impl&: This is partly just my preference to be explicit, because the execution policy only needs the allocator and nothing else from handle. Besides that it avoids introducing a dependency of ML::exec_policy that is not really needed.
Regarding the one liner: I agree that this would be nicer. @jrhemstad said that this unfortunately requires thurst 1.9.4 or newer as the thrust execution policies with older thrust versions can't be initiated with temporaries. The current workaround with auto pointers also does not allow this because the auto pointer gets out of scope before the object is used. To my understanding another nice thing with thrust 1.9.4+ for this would be that one would not need to repeat the stream. We could provide a helper function that could be used like this:

`thrust::for_each(ML::exec_policy(alloc, stream), ... );`

Reply end

Reply:
To be even more explicit I should have propably written: "Regarding passing in `ML::cumHandle_impl&` instead of `std::shared_ptr<deviceAllocator> allocator`: This is partly just my preference to be explicit, because the execution policy only needs the allocator and nothing else from handle"